adam/learner/relations.py

```diff
@@ -133,3 +155,218 @@ def _update_hypothesis(
+            # We must have at least two wildcard slots
+            return False
+        return True
+
+    def _hypotheses_from_perception(
+        self,
+        learning_state: LanguagePerceptionSemanticAlignment,
+        bound_surface_template: SurfaceTemplateBoundToSemanticNodes,
+    ) -> AbstractSet[PerceptionGraphTemplate]:
+        return immutableset(
+            PerceptionGraphTemplate.from_graph(
+                perception_graph=candidate_relation_meaning,
+                template_variable_to_matched_object_node=bound_surface_template.slot_to_semantic_node,
+            )
+            for candidate_relation_meaning in _extract_candidate_relations(
+                learning_state.perception_semantic_alignment.perception_graph,
+                bound_surface_template.slot_to_semantic_node[SLOT1],
+                bound_surface_template.slot_to_semantic_node[SLOT2],
+            )
+        )
+
+    def _can_learn_from(
+        self, language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment
+    ) -> bool:
+        # We can only learn relations if there are aligned object semantic nodes,
+        # and for there to be any aligned semantic nodes at all
+        # there must be aligned object semantic nodes,
+        # since relations, verbs, etc. build on object semantic nodes.
+        # Furthermore we only need two nodes, because relations only involve two objects.
+        return (
+            len(
+                language_perception_semantic_alignment.perception_semantic_alignment.semantic_nodes
+            )
+            > 2
+        )
+
+    def _preprocess_scene(
+        self, perception_semantic_alignment: PerceptionSemanticAlignment
+    ) -> PerceptionSemanticAlignment:
+        return perception_semantic_alignment
+
+    def log_hypotheses(self, log_output_path: Path) -> None:
+        logging.info(
+            "Logging %s hypotheses to %s",
+            len(self._concept_to_hypotheses_and_scores),
+            log_output_path,
+        )
+        for (
+            concept,
+            hypotheses_to_scores,
+        ) in self._concept_to_hypotheses_and_scores.items():
+            for (i, (hypothesis, score)) in enumerate(hypotheses_to_scores.items()):
+                hypothesis.render_to_file(
+                    concept.debug_string,
+                    log_output_path / f"{concept.debug_string}.{i}.score_{score}",
+                )
+
+        logging.info(
+            "Logging %s lexicon items to %s", len(self._lexicon), log_output_path
+        )
+        for (
+            surface_template,
+            meaning_as_perception_graph_template,
+        ) in self._lexicon.items():
+            meaning_as_perception_graph_template.render_to_file(
+                surface_template.to_short_string(),
+                log_output_path / f"lexicon-{surface_template.to_short_string()}",
+            )
+
+
+class ImmutableSet(object):
```

Comment:
Why is this here?

Reply:
Oh, pycharm import mistake with the `alt-enter` action